### PR TITLE
[Flink-12973]Support reading Hive complex types like array and map

### PR DESCRIPTION
**Pull request by zjuwangg**
## What is the purpose of the change

*Support reading Hive complex types like array and map*

## Brief change log
  - *HiveRecordSerDe support parse complex type*

## Verifying this change

This change added tests and can be verified as follows:
  - *Added test that validates complex type info can be read correctly*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by zjuwangg** [comment]
cc @bowenli86 @xuefuz @lirui-apache to review.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveRecordSerDe.java [hunk 1]
```diff
@@ -47,6 +56,12 @@ public static Object obtainFlinkRowField(Object field, ObjectInspector fieldObje
 		Object res;
 		if (fieldObjectInspector.getCategory() == ObjectInspector.Category.PRIMITIVE) {
 			res = convertPrimitiveField(field, (PrimitiveObjectInspector) fieldObjectInspector);
+		} else if (fieldObjectInspector.getCategory() == ObjectInspector.Category.STRUCT) {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveRecordSerDe.java]: maybe we can just use `HiveInspector.toFlinkObject()` and remove `obtainFlinkRowField()`?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveInputFormatTest.java [hunk 2]
```diff
@@ -122,4 +124,55 @@ public void testReadFromHiveInputFormat() throws Exception {
...
+		final String tblName = "complext_test";
+
+		TableSchema.Builder builder = new TableSchema.Builder();
+		builder.fields(new String[]{"a", "m", "s"}, new DataType[]{
+				DataTypes.ARRAY(DataTypes.INT()),
+				DataTypes.MAP(DataTypes.INT(), DataTypes.STRING()),
+				DataTypes.ROW(DataTypes.FIELD("f1", DataTypes.INT()), DataTypes.FIELD("f2", DataTypes.STRING()))});
+
+		//Now we used metaStore client to create hive table instead of using hiveCatalog for it doesn't support set
+		//serDe temporarily.
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveInputFormatTest.java]: can you open a ticket for this work?

Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveInputFormatTest.java]: ok, I created [FLINK-13061](https://issues.apache.org/jira/browse/FLINK-13061) to track it.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveRecordSerDe.java (hunk 1 above)
Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveRecordSerDe.java]: You are right. I have removed this class in the latest commit.

**Comment by bowenli86** [comment]
@zjuwangg thanks for your contribution! LGTM, will merge once CI passes

**Comment by bowenli86** [comment]
@zjuwangg there are some compile errors. Can you fix them?

**Comment by zentol** [comment]
CI report for commit 56bcca0279bde35a856dea882d8afeb43deb6938: SUCCESS [Travis](https://travis-ci.org/flink-ci/flink-ci/builds/554872195?utm_source=github_status&utm_medium=notification)

**Comment by bowenli86** [comment]
@zjuwangg Thanks for your contribution.

Merging

**Comment by bowenli86** [comment]
@zjuwangg BTW, the title should be named as `[FLINK-XXX]` rather than `[Flink-XXX]` in order to be linked back to JIRA. Please ensure the correct naming next time :)

**Pull request closed without merging**